gaplint.py

```diff
@@ -695,56 +757,270 @@ def _parse_args(kwargs):
+for rule in RULES:
+    _RULE_NAMES_AND_CODES.append([rule.name, rule.code])
+_RULE_NAMES = [x[0] for x in _RULE_NAMES_AND_CODES] + ['all']
+_RULE_CODES = [x[1] for x in _RULE_NAMES_AND_CODES]
+
+################################################################################
+# suppressions
+################################################################################
+
+def rule_code(rule):
+    '''
+    Takes a rule name or code and returns the rule code.
+    '''
+    global _RULE_NAMES, _RULE_CODES
+    if rule in _RULE_NAMES:
+        return _RULE_CODES[_RULE_NAMES.index(rule)]
+    elif rule in _RULE_CODES:
+        return rule
+    return False
+
+def make_dic(keys, vals): 
+    '''
+    Takes a list of keys and values and returns a dictionary.
+    '''
+    dic = {}
+    n = len(keys)
+    for i in xrange(n):
+        dic[keys[i]] = vals[i]
+    return dic
+
+def make_True_dic(keys):
+    '''
+    Takes a list of keys and returns a dictionary with all keys assigned the
+    value True.
+    '''
+    return make_dic(keys, [True for x in xrange(len(keys))])
+
+def filter_empty_dics(dic): # currently not used - maybe useful later
+    '''
+    Recursive function that takes a dictionary of any dimension and removes 
+    entries whose value is an empty dictionary.
+    '''
+    for key in dic.keys():
+        val = dic[key]
+        if isinstance(val, dict):
+            if val == {}:
+                del dic[key]
+            filter_no_suppression_lines(val)
+
+def is_valid_rule(rule):
+    '''
+    Takes a rule and confirms its validity, returning a bool.
+    '''
+    global _RULE_NAMES, _CODE_NAMES
+    if rule in _RULE_NAMES + _RULE_CODES:
+        return True
+    return False
+
+def get_global_suppressions(lines):
+    '''
+    Takes a list of lines and returns a dictionary with globally suppressed 
+    rules as entries, all assigned the value True.
+    '''
+    global _RULE_CODES
+    pattern1 = re.compile('\s*#')
+    pattern2 = re.compile('\s*$')
+    pattern3 = re.compile('\s*#\s*gaplint:\s*disable\s*=\s*')
+    global_supps = []
+    i = 0
+    n = len(lines)
+    myline = lines[0]
+    while  re.match(pattern1, myline) or re.match(pattern2, myline):
+        match = re.search(pattern3, myline)
+        if match:
+            pattern4 = re.compile('((\w+(-\w+)*)+(-\S+){0,1})')
+            supps = pattern4.findall(myline[match.end():])
+            for tup in supps:
+                for rule in tup:
+                    if is_valid_rule(rule):                        
+                        if rule == 'all':
+                            return make_True_dic(_RULE_CODES)
+                        code = rule_code(rule)
+                        if not code in global_supps:
+                            global_supps.append(code)        
+        if i < n - 1:
+            i += 1
+            myline = lines[i]
+    return make_True_dic(global_supps)
+
+def add_global_supps_to_line_supps(line_supp_dic, global_supp_dic):
```

Comment:
I'm not sure why you'd want to do this. This involves keeping a copy of the global suppressions for a file for every single line in the file, when only one copy is good enough. I think a different approach is required here.